### DISCUSSION TITLE
pigpio library

### DISCUSSION THREAD

**Discussion by joel5399** [rursprung/fhgr-mrproj2#51] [General]
I have developed a new package called smt_gun_package, but I'm encountering some issues with the initialization of the GPIO library.

When I launch only the smt_gun_controller, everything works fine. However, if the movement_controller is launched beforehand, I encounter the same errors as usual, such as 'can't bind to port 8889'.

I thought about checking if the movement node is running at the beginning and, if it is, skipping the gpioInitialise() function. However, if I attempt to write to a GPIO, I receive an error stating that the library is not initialized.

Do you have any ideas on how I could resolve this problem? My only thought is to create a separate package that specifically handles the GPIOs and subscribes to a GPIO topic.

**Comment by rursprung** [comment]
it seems that the way we're currently using the library it can only be used by one concurrent process, which is of course an issue. i briefly googled and [found this answer](https://raspberrypi.stackexchange.com/a/65981) which gives various options.

we briefly discussed the following two options:
* merge `smt_gun_controller` and `smt_movement_controller` into a single package so that we only have one process which needs to interact with the GPIO
  * advantage: simple solution
  * disadvantage: bad project structuring as the two are not related to each other
* create new `smt_rpi_gpio_controller` (or similarly named) which can be called by the other two controllers to interact with the GPIO pins (calls can probably best be done using ROS service calls to avoid the performance lag of ROS messages)
  * advantage: clean separation between the gun & movement controllers
  * disadvantage: more complex setup

what we haven't discussed (because i wasn't aware of it) was the third option: use `pigpiod_if2` and just launch `pigpiod` separately. i think that would be the cleanest solution, would probably require little changes to the existing code and would also solve #46.

**Comment by joel5399** [comment]
Currently, I'm trying to implement the second option. I'm already quite far along with it. At the moment, I just need to test it on the hardware. If it works, we can leave everything as it is for now. If we later realize that the third solution would be better and cause fewer problems, we can still implement that one later.

Is this okay for you?

**Reply by rursprung** [reply]
fine with me!
